### ISSUE TITLE
Training Seeker on WOI data clarifications as in paper

### ISSUE THREAD

**Issue by tiger241** [facebookresearch/ParlAI#4935]
Hi I am new at this coding. I have done some playing around with the seeker code based on this.
My goal is to reproduce the paper results on WOI while going through the code in detail.

https://parl.ai/projects/seeker/
$ python -m parlai.scripts.multiprocessing_train \
--task projects.seeker.tasks.knowledge,projects.seeker.tasks.dialogue,projects.seeker.tasks.search_query \
--multitask-weights 2,2,1 -bs 1 -vstep 1000 -vmt ppl -vp 5 -vmm min -vme 100000 -lstep 50 \
--init-opt arch/r2c2_base_3B --init-model zoo:seeker/r2c2_blenderbot_3B/model \
--model projects.seeker.agents.seeker:ComboFidGoldDocumentAgent --n-docs 5 \
--text-truncate 1000 --label-truncate 128 --truncate 1000 \
--fp16 True -lr 1e-06 --lr-scheduler reduceonplateau --optimizer adamw --save-after-valid True \
--warmup-updates 100 --update-freq 1 --gradient-clip 1.0 --skip-generation True --dropout 0.1 \
--attention-dropout 0.0 --load-from-checkpoint true --ddp-backend zero2 \
--checkpoint-activations true --model-file /tmp/my_seeker_dialogue_model

But switched multiprocessing_train to train_model. I am looking into understanding the code before running multiprocessing_train, so I have the code in pycharm and going through each line. 

$ python generate_lm_data.py --task my_task --save-dir path/to/savedir --search-server <server> --datatype train
1) **_The documentation does not clarify the task I should use, so I am confused_**
_**I set task=projects.seeker.tasks.search_query:SearchQueryTeacher **_
**_Not sure what to put it in here to produce what is required for the training of seeker_**

I understand how to make the server but i have 2 more questions. 

2) Do you guys have a retreiver model available? I can make a quick one myself (TF-IDF to save computation) but you must have at least some bert ranker based on my navigation of the code ! I am just not sure how to call it in the ParlAI code.

3) This might be a novice question:
 When I run the generate lm code I get the train, validation and test .jsonl data, 
but I am not how to set up the retriever on that data.

The first act: 
{'id':  'WoiSearchQueryTeacher',
'is_last_search_query': True,
'skip_retrieval': True,
'text': 'I live in New York, New York.\nI like going to the moives.\n I love to get the overpriced snacks. __generate-query__',
'episode_done': True,
'eval_labels': ['upcoming movies'] }

Search_queries =   ['upcoming movies'] 

searchagent should ideally retrieve the respective golden document contents for the conversation with text field 'upcoming movies' is that correct?

So in my case I can make TF-IDF over the docs in the jsonl files and the wrap it in a server.
Alternative I can give a golden retreival by investigating the text fields and finding the matching docs (golden knowledge retreiver)

**Comment by klshuster** [comment]
Thanks for your interest in SeeKeR! In order:

1. Training a SeeKeR dialogue model does not require any LM data; so, you can ignore this step if that is your goal. In fact, you can just use the data provided as-is (within `projects/seeker/tasks`). If you wanted to train a SeeKeR language model, then you would want to use that script. The data should be language-modeling data; in the paper we used data crawled from the internet.
2. Yes! We provide several different retrievers via the `--rag-retriever-type` argument (in the full SeeKeR setup, this would be the `krm_rag_retriever_type`); I would take a look at `agents/rag/retrievers.py` for more information 
3. Yes, you can use TFIDF over documents, where the input would be `'upcoming movies'`

**Comment by tiger241** [comment]
Thank you!! I think I understand what to do for the retrievers!
1) However, I am a bit confused about the terminology. 
My understanding is that the dialogue task would do all the training (unless you mean deploying everything)
(at least in the paper in the dialogue task you guys list all the tasks and what you do)

2) I would like to train a model for knowledge and response but I want to make some customized changes to the Search task.
So ideally I will take the code with the 2 tasks and then plug in my version of Search (let's assume I want to use gpt2 medium like in the example on the web page)

I noticed you have knowledge and dialogue in that task folder and they seem to include all the data tasks.
If I run task=projects.seeker.tasks.knowledge then it will train through all the datasets for the knowledge component does that make sense?

I am confused about the difference between lm:KnowledgeTeacher and knowledge.KnowledgeTeacher

Here is my thought process:
  So I'll go with the first two:
  projects.seeker.tasks.lm:KnowledgeTeacher,
  projects.seeker.tasks.ResponseTeacher,
  projects.seeker.tasks.SearchQueryTeacher (not needed as I will plug in my version of search)

I would like to go through the training code in detail to make sure I can understand it well, but do you guys have the model weights available for BART and GPT2?

Many thanks!!

**Comment by klshuster** [comment]
Sorry, I see where the confusion is arising.

`seeker_dialogue` --> A model trained for *dialogue* applications, where the S(earch), K(nowledge), and R(esponse) modules are trained with *dialogue*-oriented tasks. This corresponds to `projects.seeker.tasks.search_query:SearchQueryTeacher,projects.seeker.tasks.knowledge:KnowledgeTeacher,projects.seeker.tasks.dialogue:DialogueTeacher`. The knowledge tasks here are built from dialogue (e.g., Wizard of Wikipedia) and question-answering (e.g., SQuAD) datasets.

`seeker_lm` --> A model trained for *language modeling* applications, where the S(earch), K(nowledge), and R(esponse) modules are trained with *language modeling*-oriented tasks. This corresponds to the teachers in the `projects.seeker.tasks.lm` file. You need to build the knowledge datasets yourself, but they would derive from language modeling datasets.

So, you should probably use `projects.seeker.tasks.knowledge:KnowledgeTeacher` and `projects.seeker.tasks.knowledge:ResponseTeacher`, with your custom search query teacher.

We offer the raw GPT2 and BART weights through parlai (`--agent hugging_face/gpt2` or `--agent bart`), but neither trained on the SeeKeR tasks. Alternatively, we do have a `--model-file zoo:seeker/seeker_400M_dialogue/model` that is a BART-style model trained on the seeker dialogue tasks.

**Comment by klshuster** [comment]
looks good to me!
